src/ray/raylet/worker_pool.cc

```diff
@@ -899,14 +935,30 @@ std::shared_ptr<WorkerInterface> WorkerPool::PopWorker(
     if (worker == nullptr) {
       // There are no more non-actor workers available to execute this task.
       // Start a new worker process.
-      proc = StartWorkerProcess(task_spec.GetLanguage(), rpc::WorkerType::WORKER,
-                                task_spec.JobId(), {}, /* dynamic_options */
-                                task_spec.SerializedRuntimeEnv(),
-                                task_spec.OverrideEnvironmentVariables());
+
+      if (!(task_spec.SerializedRuntimeEnv() == "{}" ||
+            task_spec.SerializedRuntimeEnv() == "")) {
+        // create runtime env.
+        create_runtime_env = true;
+        agent_manager_->CreateRuntimeEnv(
+            task_spec.SerializedRuntimeEnv(),
+            [start_worker_process_fn, &state, task_spec](bool done) {
+              if (!done) {
+                // TODO(guyang.sgy): Reschedule to other nodes when create runtime env
+                // failed.
+                RAY_LOG(ERROR) << "Create runtime env rpc failed. Wait for next time to "
+                                  "retry or reschedule.";
+                return;
```

Comment:
Why don't we need to add to and erase from `state.tasks_to_pending_runtime_envs` like we did in this part? https://github.com/ray-project/ray/pull/15904/files#diff-4c19d6ebd8575189e9439fa5646b567546dc54f64d576a959ce1bb43902d1a03R882-R887

Reply:
Because this worker is not a dedicated worker which bound to a task. If we put it to `tasks_to_pending_runtime_envs`, it could not be reused.